File: .claude/hooks/guard-push-main.mjs
#!/usr/bin/env node
// @hookstack pre-bash-guard-git-push-main
// Bloque git push --force vers main/master (PreToolUse Bash)
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';

export function run(input) {
  const command = input.tool_input?.command ?? '';
  const isForce =
    /git\s+push\b.*--force(?:-with-lease)?/.test(command) ||
    /git\s+push\b.*-f\b/.test(command);
  const isMain = /\b(main|master)\b/.test(command);

  return isForce && isMain
    ? {
        decision: 'block',
        reason: 'Force-push vers main/master interdit. Créez une PR ou demandez confirmation explicite.',
      }
    : null;
}

/* v8 ignore next 5 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const input = JSON.parse(readFileSync(0, 'utf8'));
  const result = run(input);
  if (result) process.stdout.write(JSON.stringify(result));
}

File: .claude/hooks/guard-push-main.py
#!/usr/bin/env python3
# @hookstack pre-bash-guard-git-push-main
"""Blocks git push --force to main/master (PreToolUse Bash)."""
import json
import re
import sys

FORCE_RE = re.compile(r"git\s+push\b.*--force(?:-with-lease)?|git\s+push\b.*-f\b")
MAIN_RE = re.compile(r"\b(main|master)\b")


def run(input_data):
    command = (input_data.get("tool_input") or {}).get("command") or ""
    is_force = bool(FORCE_RE.search(command))
    is_main = bool(MAIN_RE.search(command))

    if is_force and is_main:
        return {
            "decision": "block",
            "reason": "Force-push vers main/master interdit. Créez une PR ou demandez confirmation explicite.",
        }
    return None


if __name__ == "__main__":
    data = json.load(sys.stdin)
    result = run(data)
    if result:
        sys.stdout.write(json.dumps(result))
